### PR TITLE
[WIP] alloc_pages

### PR DESCRIPTION
**Pull request by misson20000**
This PR includes some pretty major refactors and reworks of ace_loader.

First of all, `ace.nro` is now `wk_ace.nro`, since it's specific to webkit. There's another binary now too, `sysmod_ace.nro`. This is pretty much the same thing, but without the webkit-specific cleanup, making it suitable for using under sysmodules.

Next, the whole memory mapping system has been ripped out and replaced with `alloc_pages`. The allocator probably isn't very fast, and isn't very cool and good or anything. This cleans up some of the memory management aspects of ace_loader, making it harder to accidentally overwrite something important with an NRO that we're downloading or something. Something to be mindful of at the moment is that if ace_loader's parent config didn't override the heap, I'm just kinda making an (educated) guess as to how much heap space we're using and that shouldn't get added to the alloc_pages arena and overwriting loader_config's values with it. This is really horrible and we might want to do something better, but we're also always overriding heap on hardware and I just did it so I could test under Mephisto.

I haven't changed HTTP loading of NROs yet to use alloc_pages, which is one of the things that needs to be done before this gets merged.

To finish the job for alloc_pages, we need to add a Homebrew ABI key, add bindings in libtransistor, and implement sbrk as a loop of `map(continuous heap region, alloc_pages(min = 0x1000, max = incr))`.

Both `wk_ace` and `sysmod_ace` are working for me with `test_helloworld.nro`. Not sure if `sysmod_ace` would work under WebKit.

### CODE REVIEW

File: projects/ace_loader/common_memory.c [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+
+	if(free_head->is_locked) {
+		return true;
+	}
+	
+	if(free_head->is_free) {
+		return true;
+	}
+
+	// TODO: check memory permission/type/state
```
Comment (misson20000) [inline projects/ace_loader/common_memory.c]: TODO: check memory permission/type/state

File: projects/ace_loader/common_setup.c [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+		uint64_t r;
+		printf("- starting autorun\n");
+		nro_arg_name("autorun");
+		nro_add_arg(http_hostname);
+		r = nro_execute(heap_base, ret);
+		printf("- NRO returned 0x%016lX\n", r);
+		// do not exit for GUIs
+		if(!nro_loaded_count || nro_unload_fail)
+			return 0;
+			}*/
```
Comment (misson20000) [inline projects/ace_loader/common_setup.c:127]: fix this

File: projects/ace_loader/http.c [hunk 3]
```diff
@@ -154,6 +154,7 @@ int http_get_file(const char *path, void *buff, uint64_t maxsize)
 
 	// get an answer
 	ret = parse_header(temp, sizeof(temp), &offs, &got);
+	printf("- parse_header > %d\n", ret);
```
Comment (misson20000) [inline projects/ace_loader/http.c:157]: remove this

File: projects/ace_loader/memory.c [hunk 4]
```diff
@@ -1,235 +0,0 @@
...
-		if(minfo.permission == 3 && minfo.memory_type == 5 && minfo.memory_attribute & 8)
-			printf("- svcSetMemoryAttribute 0x%06X\n", svcSetMemoryAttribute(minfo.base_addr, minfo.size, 0, 0));
-
-		addr = minfo.base_addr + minfo.size;
-		if(!addr)
-			break;
-	}
-
-	return 0;
-}
```
Comment (misson20000) [inline projects/ace_loader/memory.c:170]: for other reviewers: this became `wk_mem_heap_cleanup()`

File: projects/ace_loader/memory.c [hunk 5]
```diff
@@ -1,235 +0,0 @@
...
-		}
-		addr = minfo.base_addr + minfo.size;
-		if(!addr)
-			break;
-	}
-
-	if(!ptr || !wkBase)
-		return 1;
-	return 0;
-}
```
Comment (misson20000) [inline projects/ace_loader/memory.c:234]: becomes `wk_mem_install_hook`

File: projects/ace_loader/nro.c [hunk 6]
```diff
@@ -149,35 +152,45 @@ result_t nro_load(void *load_base, int in_size)
...
+		return 3;
+	}
+	
 	// clear bss
-	memset(load_base + nro_size, 0, bss_size);
+	memset(nro_bss, 0, bss_size);
 
 	// load NRO
-	r = ro_load_nro(&nro_base, load_base, nro_size, load_base + nro_size, bss_size);
+	r = ro_load_nro(&nro_base, nro_blob, nro_size, nro_bss, bss_size);
```
Comment (misson20000) [inline projects/ace_loader/nro.c:182]: the `nro_base` and `nro_load_base` terminology is really confusing

File: projects/ace_loader/server.c [hunk 7]
```diff
@@ -80,17 +85,20 @@ int server_init()
 	};
 
 	sck = bsd_socket(2, 1, 6); // AF_INET, SOCK_STREAM, PROTO_TCP
+	printf("- server socket: %d, 0x%x, %d\n", sck, bsd_result, bsd_errno);
```
Comment (misson20000) [inline projects/ace_loader/server.c:85]: remove this

File: projects/ace_loader/server.c [hunk 8]
```diff
@@ -432,12 +479,12 @@ void func_exec(char *par)
 		r = nro_execute(heap_base, ret);
 		printf("- NRO returned 0x%016lX\n", r);
 	} else
-		printf("- get NRO error %i\n", -ret);
+	printf("- get NRO error %i\n", -ret);*/
```
Comment (misson20000) [inline projects/ace_loader/server.c]: fix this

File: projects/ace_loader/sysmod_main.c [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+		}
+
+		/*ret = pm_init();
+		if(ret != 0) {
+			return ret;
+		}
+		printf("- terminating browser...\n");
+		ret = pm_terminate_process_by_title_id(0x0100000000001011);
+		printf("-   => 0x%x\n", ret);
+		pm_finalize();*/
```
Comment (misson20000) [inline projects/ace_loader/sysmod_main.c]: clean this up

File: projects/ace_loader/sysmod_main.c [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+		if(ret != 0) {
+			return ret;
+		}
+
+		if(!loader_config.heap_overridden) {
+			// hack to keep us from tripping over our own heap
+			loader_config.heap_overridden = true;
+			loader_config.heap_base = ptr;
+			loader_config.heap_size = (malloc(1) - ptr) + 0x8000;
+		}
```
Comment (misson20000) [inline projects/ace_loader/sysmod_main.c]: this hack is terrible

File: projects/ace_loader/sysmod_main.c [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+		
+		ret = mem_init();
+		if(ret != 0) {
+			return ret;
+		}
```
Comment (misson20000) [inline projects/ace_loader/sysmod_main.c]: get main thread handle?

**Comment by misson20000** [comment]
mentioning #102 and #69

**Comment by kgsws** [comment]
I do not see any obvious problems.
Also i suspect HTTP 'module' might become obsolete with SD card access.

**Pull request merged**